sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala

```diff
@@ -290,29 +291,88 @@ case class DescribeTableCommand(table: TableIdentifier, isExtended: Boolean)
-        relation.schema.fields.foreach { field =>
-          val comment =
-            if (field.metadata.contains("comment")) field.metadata.getString("comment") else ""
-          result += Row(field.name, field.dataType.simpleString, comment)
-        }
+        describeSchema(relation.schema, result)
     }
 
     result
   }
+
+  // Shows data columns and partitioned columns (if any)
+  private def describe(relation: CatalogRelation, buffer: ArrayBuffer[Row]): Unit = {
+    describeSchema(relation.catalogTable.schema, buffer)
+
+    if (relation.catalogTable.partitionColumns.nonEmpty) {
+      buffer += Row("# Partition Information", "", "")
+      buffer += Row(s"# ${output(0).name}", output(1).name, output(2).name)
+
+      describeSchema(relation.catalogTable.partitionColumns, buffer)
+    }
+  }
+
+  private def describeExtended(relation: CatalogRelation, buffer: ArrayBuffer[Row]): Unit = {
+    describe(relation, buffer)
+
+    buffer += Row("", "", "")
+    buffer += Row("# Detailed Table Information", relation.catalogTable.toString, "")
+  }
+
+  private def describeFormatted(relation: CatalogRelation, buffer: ArrayBuffer[Row]): Unit = {
+    describe(relation, buffer)
+
+    val table = relation.catalogTable
+
+    buffer += Row("", "", "")
+    buffer += Row("# Detailed Table Information", "", "")
+    buffer += Row("Database:", table.database, "")
+    buffer += Row("Owner:", table.owner, "")
+    buffer += Row("Create Time:", new Date(table.createTime).toString, "")
+    buffer += Row("Last Access Time:", new Date(table.lastAccessTime).toString, "")
+    buffer += Row("Location:", table.storage.locationUri.getOrElse(""), "")
+    buffer += Row("Table Type:", table.tableType.name, "")
+
+    buffer += Row("Table Parameters:", "", "")
+    table.properties.foreach { case (key, value) =>
+      buffer += Row(s"  $key", value, "")
+    }
+
+    buffer += Row("", "", "")
+    buffer += Row("# Storage Information", "", "")
+    table.storage.serde.foreach(serdeLib => buffer += Row("SerDe Library:", serdeLib, ""))
+    table.storage.inputFormat.foreach(format => buffer += Row("InputFormat:", format, ""))
+    table.storage.outputFormat.foreach(format => buffer += Row("OutputFormat:", format, ""))
+    buffer += Row("Compressed:", if (table.storage.compressed) "Yes" else "No", "")
+    buffer += Row("Num Buckets:", table.numBuckets.toString, "")
+    buffer += Row("Bucket Columns:", table.bucketColumnNames.mkString("[", ", ", "]"), "")
+    buffer += Row("Sort Columns:", table.sortColumnNames.mkString("[", ", ", "]"), "")
+
+    buffer += Row("Storage Desc Parameters:", "", "")
+    table.storage.serdeProperties.foreach { case (key, value) =>
+      buffer += Row(s"  $key", value, "")
+    }
+  }
+
+  private def describeSchema(schema: StructType, buffer: ArrayBuffer[Row]): Unit = {
+    schema.foreach { column =>
+      val comment =
+        if (column.metadata.contains("comment")) column.metadata.getString("comment") else ""
```

Comment:
I think the metadata is heavily used in ML code. Another thing is that the data type API is already public for a long time. We probably don't want to change it unless there are particular good reasons.